### PR TITLE
Support for joins

### PR DESCRIPTION
**Pull request by morganchen12**
Mirrors https://github.com/firebase/FirebaseUI-Android/pull/276.

~~Still missing a few things:~~
- [x] ~~Documentation~~
- [x] ~~Actual error handling~~
- [x] ~~Data source tests~~
- [x] ~~Tests for edge cases~~

Good to merge, pending some code review.

cc @puf, @ianbarber

### CODE REVIEW

**Comment by morganchen12** [comment]
Seems like github is bugging out and not updating this PR correctly.

File: FirebaseDatabaseUI/FirebaseIndexArray.h [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+- (void)array:(FirebaseIndexArray *)array
+    reference:(FIRDatabaseReference *)ref
+      atIndex:(NSUInteger)index
+didFailLoadWithError:(NSError *)error;
+
+/**
+ * Delegate method which is called whenever an object is added to a
+ * FirebaseArray. On a
+ * FirebaseArray synchronized to a Firebase reference, this corresponds to an
+ * [FIRDataEventTypeChildAdded](https://www.firebase.com/docs/ios/guide/retrieving-data.html#section-event-types)
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.h]: Please replace these doc links with ones to firebase.google.com across the file

File: FirebaseDatabaseUI/FirebaseIndexArray.h [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ * event being raised.
+ * @param object The database reference that has moved locations
+ * @param fromIndex The index the reference is being moved from
+ * @param toIndex The index the reference is being moved to
+ */
+- (void)array:(FirebaseIndexArray *)array didMoveReference:(FIRDatabaseReference *)ref fromIndex:(NSUInteger)fromIndex toIndex:(NSUInteger)toIndex;
+
+/**
+ * Delegate method which is called whenever the backing query is canceled. This error is fatal
+ * and the index array will become unusable afterward, so please handle it appropriately.
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.h]: What is the appropriate handling? Guessing effectively to stop using this object

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+//
+
+// clang-format on
+
+#import "FirebaseIndexArray.h"
+#import "FirebaseQueryObserver.h"
+
+@interface FirebaseIndexArray () <FirebaseArrayDelegate>
+
+@property (nonatomic, readonly) id<FIRDataObservable> index;
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: Not a big fan of the readonly here - if we're declaring private property decls lets make them readwrite and use self.property when setting.

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+- (instancetype)initWithIndex:(id<FIRDataObservable>)index
+                         data:(id<FIRDataObservable>)data
+                     delegate:(nullable id<FirebaseIndexArrayDelegate>)delegate; {
+  NSParameterAssert(index != nil);
+  NSParameterAssert(data != nil);
+  self = [super init];
+  if (self != nil) {
+    _index = index;
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: as with comment above, would prefer self.index

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  return self;
+}
+
+- (instancetype)initWithIndex:(id<FIRDataObservable>)index
+                         data:(id<FIRDataObservable>)data {
+  return [self initWithIndex:index data:data delegate:nil];
+}
+
+- (void)observeQueries {
+  _indexArray = [[FirebaseArray alloc] initWithQuery:self.index delegate:self];
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: again would refer self.indexArray

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  for (FirebaseQueryObserver *observer in observers) {
+    if (observer.contents != nil) {
+      [array addObject:observer.contents];
+    }
+  }
+  return [array copy];
+}
+
+- (NSUInteger)count {
+  return self.observers.count;
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: while  here _observers.count would be fine

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+- (void)dealloc {
+  [self invalidate];
+}
+
+#pragma mark - FirebaseArrayDelegate
+
+// These delegate methods are entirely responsible
+// for keeping a local array of queries up to date.
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: This comment is a bit abandoned here. I might be inclined to move it into a top level class comment that explains what we are trying to do and how we go about it.

File: FirebaseDatabaseUI/FirebaseIndexArray.h [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+ * @param error the error that was raised
+ */
+- (void)array:(FirebaseIndexArray *)array queryCancelledWithError:(NSError *)error;
+
+@end
+
+@interface FirebaseIndexArray : NSObject
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.h]: This needs a decent class comment explaining what this is for and how it should be used

File: FirebaseDatabaseUI/FirebaseIndexArray.m (hunk 3 above)
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: Note: discussed this separately, so either way works as long as consistent.

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+@property (nonatomic, readonly) id<FIRDataObservable> index;
+@property (nonatomic, readonly) id<FIRDataObservable> data;
+
+@property (nonatomic, readonly) FirebaseArray *indexArray;
+
+@property (nonatomic, readonly) NSMutableArray<FirebaseQueryObserver *> *observers;
+
+@end
+
+@implementation FirebaseIndexArray
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: Add class comment about any implementation level details (see comment below)

File: FirebaseDatabaseUI/FirebaseIndexArray.m [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    [NSException exceptionWithName:@"FIRUnavailableMethodException"
+                            reason:@"-init is unavailable. Please use the designated initializer instead."
+                          userInfo:nil];
+  @throw e;
+}
+
+- (instancetype)initWithIndex:(id<FIRDataObservable>)index
+                         data:(id<FIRDataObservable>)data
+                     delegate:(nullable id<FirebaseIndexArrayDelegate>)delegate; {
+  NSParameterAssert(index != nil);
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: As discussed, see whether replacing one firebasearray with a query makes the code easier.

Comment (morganchen12) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: I tried this and ended up duplicating much of the `index` logic since queries only give you the `previousChildKey`. Might be worth extracting that from FirebaseArray into a more inert data object, but for now I think keeping FirebaseArray here is cleaner, if not less verbose.

**Comment by morganchen12** [comment]
@ianbarber ready for more review

File: FirebaseDatabaseUI/FirebaseArrayDelegate.h [hunk 11]
```diff
@@ -22,52 +22,43 @@
...
- * event being raised.
+ * FirebaseArray. On a FirebaseArray synchronized to a Firebase reference, 
+ * this corresponds to a @c FIRDataEventTypeChildAdded event being raised.
  * @param object The object added to the FirebaseArray
  * @param index The index the child was added at
  */
 - (void)array:(FirebaseArray *)array didAddObject:(id)object atIndex:(NSUInteger)index;
 
 /**
  * Delegate method which is called whenever an object is chinged in a
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseArrayDelegate.h]: Not your fault but chinged -> changed

File: FirebaseDatabaseUI/FirebaseIndexArray.h [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+/**
+ * Returns the number of items in the array.
+ */
+@property(nonatomic, readonly) NSUInteger count;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+/**
+ * Initializes a FirebaseIndexArray with an index query and a data query.
+ * The array expects the keys of the children of the index query to be children
```
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.h]: (here and elsewhere) change "to be" to "to match" - this is complex stuff, and clarifying that we have the same values in two places is probably pertinent.

File: FirebaseDatabaseUI/FirebaseIndexArray.m (hunk 10 above)
Comment (ianbarber) [inline FirebaseDatabaseUI/FirebaseIndexArray.m]: ack
